### PR TITLE
Feature/trm 29297 support ttl and nt formats

### PR DESCRIPTION
**Pull request by supun-ebi**

### CODE REVIEW

File: common-rest/src/main/java/org/uniprot/api/rest/service/NTPrologs.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package org.uniprot.api.rest.service;
+
+public class NTPrologs {
+    private NTPrologs() {}
+
```
Comment (ahmadshadab) [inline common-rest/src/main/java/org/uniprot/api/rest/service/NTPrologs.java:5]: NTriples instead of NT is more clearer

File: common-rest/src/test/java/org/uniprot/api/common/repository/stream/rdf/RDFStreamerTest.java [hunk 2]
```diff
@@ -191,4 +93,4 @@ void idsToRDFTupleStoreStream() {
                         + "</rdf:RDF>",
                 rdfString);
     }
```
Comment (ahmadshadab) [inline common-rest/src/test/java/org/uniprot/api/common/repository/stream/rdf/RDFStreamerTest.java:95]: add tests for multiple ids and empty response

File: common-rest/src/main/java/org/uniprot/api/rest/controller/BasicSearchController.java [hunk 3]
```diff
@@ -271,6 +274,14 @@ protected boolean isRDFAccept(HttpServletRequest request) {
         return contentType.equals(RDF_MEDIA_TYPE);
     }
 
+    protected Optional<String> getAcceptedCustomType(HttpServletRequest request) {
+        MediaType contentType = getAcceptHeader(request);
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/controller/BasicSearchController.java:278]: Done

File: common-rest/src/main/java/org/uniprot/api/rest/service/RDFService.java [hunk 4]
```diff
@@ -1,66 +1,63 @@
...
+        this.tagProvider = tagProvider;
         this.restTemplate = restTemplate;
         this.clazz = clazz;
+        this.type = type;
+        this.format = format;
     }
 
     @Override
     public List<T> getEntries(Iterable<String> accessions) {
         List<String> allAccessions = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/service/RDFService.java:32]: done

File: common-rest/src/main/java/org/uniprot/api/rest/service/TTLPrologs.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                    + "@prefix foaf: <http://xmlns.com/foaf/0.1/> .\n"
+                    + "@prefix owl: <http://www.w3.org/2002/07/owl#> .\n"
+                    + "@prefix pubmed: <http://purl.uniprot.org/pubmed/> .\n"
+                    + "@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n"
+                    + "@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .\n"
+                    + "@prefix skos: <http://www.w3.org/2004/02/skos/core#> .\n"
+                    + "@prefix up: <http://purl.uniprot.org/core/> .\n"
+                    + "@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .";
+
+    private TTLPrologs() {}
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/service/TTLPrologs.java:124]: agree, will mention in a jira and add to the backlog

File: common-rest/src/main/java/org/uniprot/api/rest/service/RDFService.java [hunk 6]
```diff
@@ -82,17 +79,16 @@ private T getEntriesByAccessions(List<String> accessions) {
      * beginning of stream and RDF_CLOSE_TAG will be added in the end of the stream. see
      * Stream.concat in StoreStreamer
      *
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/service/RDFService.java:81]: done

File: common-rest/src/main/java/org/uniprot/api/rest/service/TagProvider.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public int getEndingPosition(String body, String format) {
+        switch (format) {
+            case RDF:
+                return body.indexOf("</rdf:RDF>");
+            case TTL:
+            case NT:
+                return body.length();
+        }
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/service/TagProvider.java:32]: No, because it has a closing tag while TTL and NT do not.

File: common-rest/src/main/java/org/uniprot/api/rest/service/RDFService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/service/RDFService.java:81]: done

File: common-rest/src/main/java/org/uniprot/api/rest/service/NTPrologs.java (hunk 1 above)
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/service/NTPrologs.java:5]: done

File: common-rest/src/main/java/org/uniprot/api/rest/output/converter/NTMessageConverter.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package org.uniprot.api.rest.output.converter;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+import org.uniprot.api.common.concurrency.Gatekeeper;
+import org.uniprot.api.rest.output.UniProtMediaType;
+
+public class NTMessageConverter extends AbstractEntityIdHttpMessageConverter<String> {
+    public NTMessageConverter() {
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/output/converter/NTMessageConverter.java:10]: done

File: common-rest/src/main/java/org/uniprot/api/rest/output/UniProtMediaType.java [hunk 9]
```diff
@@ -31,9 +31,12 @@ public class UniProtMediaType {
...
     public static final String RDF_MEDIA_TYPE_VALUE = "application/rdf+xml";
     public static final MediaType RDF_MEDIA_TYPE = valueOf(RDF_MEDIA_TYPE_VALUE);
-
     public static final String HDF5_MEDIA_TYPE_VALUE = "application/x-hdf5";
     public static final MediaType HDF5_MEDIA_TYPE = valueOf(HDF5_MEDIA_TYPE_VALUE);
+    public static final String TTL_MEDIA_TYPE_VALUE = "application/ttl+xml";
+    public static final MediaType TTL_MEDIA_TYPE = valueOf(TTL_MEDIA_TYPE_VALUE);
+    public static final String NT_MEDIA_TYPE_VALUE = "application/nt+xml";
+    public static final MediaType NT_MEDIA_TYPE = valueOf(NT_MEDIA_TYPE_VALUE);
     public static final String MARKDOWN_MEDIA_TYPE_VALUE = TEXT_MARKDOWN_VALUE;
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/rest/output/UniProtMediaType.java:40]: done

File: common-rest/src/main/java/org/uniprot/api/common/repository/stream/rdf/PrologProvider.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        switch (format) {
+            case RDF:
+                return getForRdf(type);
+            case TTL:
+                return getForTtl(type);
+            case NT:
+                return NTPrologs.NT_COMMON_PROLOG;
+        }
+        throw new IllegalArgumentException(String.format("Unsupported format %s", format));
+    }
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/main/java/org/uniprot/api/common/repository/stream/rdf/PrologProvider.java:34]: done

File: common-rest/src/test/java/org/uniprot/api/common/repository/stream/rdf/RDFStreamerTest.java (hunk 2 above)
Comment (supun-ebi) [reply common-rest/src/test/java/org/uniprot/api/common/repository/stream/rdf/RDFStreamerTest.java:95]: done

File: uniprotkb-rest/src/main/java/org/uniprot/api/uniprotkb/service/UniProtEntryService.java [hunk 11]
```diff
@@ -217,9 +217,10 @@ public String findAccessionByProteinId(String proteinId) {
         }
     }
 
-    public Stream<String> streamRDF(UniProtKBStreamRequest streamRequest) {
+    public Stream<String> streamRDF(
+            UniProtKBStreamRequest streamRequest, String type, String format) {
         SolrRequest solrRequest = createDownloadSolrRequest(streamRequest);
-        return this.uniProtRDFStreamer.idsToRDFStoreStream(solrRequest);
+        return rdfStreamer.stream(documentIdStream.fetchIds(solrRequest), type, format);
     }
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply uniprotkb-rest/src/main/java/org/uniprot/api/uniprotkb/service/UniProtEntryService.java:224]: done

File: common-rest/src/test/java/org/uniprot/api/common/repository/stream/rdf/RDFStreamerTest.java [hunk 12]
```diff
@@ -191,4 +93,4 @@ void idsToRDFTupleStoreStream() {
...
-}
+}
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply common-rest/src/test/java/org/uniprot/api/common/repository/stream/rdf/RDFStreamerTest.java:96]: done

File: uniprotkb-rest/src/main/java/org/uniprot/api/uniprotkb/service/UniProtEntryService.java [hunk 13]
```diff
@@ -86,7 +84,8 @@ public UniProtEntryService(
             FacetTupleStreamTemplate facetTupleStreamTemplate,
             UniProtQueryProcessorConfig uniProtKBQueryProcessorConfig,
             SearchFieldConfig uniProtKBSearchFieldConfig,
-            RDFStreamer uniProtRDFStreamer) {
+            TupleStreamDocumentIdStream documentIdStream,
+            RDFStreamer uniprotRdfStreamer) {
         super(
```
[the remark this answers is not in the document]
Comment (supun-ebi) [reply uniprotkb-rest/src/main/java/org/uniprot/api/uniprotkb/service/UniProtEntryService.java:89]: done

**Review by ahmadshadab — APPROVED** [review]
Good job man. It has a lot of changes. Please split the task in future to avoid so many changes in one PR.
Merge after you add two types in CSVSOurce of method checkMediaTypeFileExtensionRoundTrip
